Fixes for disassembly support in x86-hosted R2RDump

trylek opened the pull request:
1) We need to remove the PlatformTarget spec to make sure
x86-hosted R2RDump actually runs under 32-bit host, otherwise it
fails to load x86-targeting coredistools.dll. Thanks to Michal
for pointing that out.

2) I have removed Zach's workaround blocking disassembly on x86.

3) Based on JanV's advice I have marked all the interface methods
as using the cdecl calling convention in accordance with the
native declarations.

4) I found out that Address was incorrectly declared as a 64-bit
integer in the PInvoke declaration while in reality it should be
an IntPtr (native int).

With these changes I'm for the first time able to disassemble
x86-targeting System.Private.CoreLib using the x86-hosted R2RDump.

Thanks

Tomas

/cc @dotnet/crossgen-contrib 

Fixes: https://github.com/dotnet/runtime/issues/45857
Fixes: https://github.com/dotnet/runtime/issues/10928

trylek committed:
Fixes for disassembly support in x86-hosted R2RDump

1) We need to remove the PlatformTarget spec to make sure
x86-hosted R2RDump actually runs under 32-bit host, otherwise it
fails to load x86-targeting coredistools.dll. Thanks to Michal
for pointing that out.

2) I have removed Zach's workaround blocking disassembly on x86.

3) Based on JanV's advice I have marked all the interface methods
as using the cdecl calling convention in accordance with the
native declarations.

4) I found out that Address was incorrectly declared as a 64-bit
integer in the PInvoke declaration while in reality it should be
an IntPtr (native int).

With these changes I'm for the first time able to disassemble
x86-targeting System.Private.CoreLib using the x86-hosted R2RDump.

Thanks

Tomas

jkotas commented on src/coreclr/tools/r2rdump/CoreDisTools.cs:
```diff
@@ -24,23 +24,23 @@ public enum TargetArch
 
-        [DllImport(_dll)]
+        [DllImport(_dll, CallingConvention = CallingConvention.Cdecl)]
         [return: MarshalAs(UnmanagedType.I4)]
```
```suggestion
```
Unnecessary

trylek replied:
Great, thanks for confirming, I was also wondering what's the use of this clause ;-).

trylek replied:
Removed in 2nd commit.

trylek committed:
Remove superfluous return type marshalling attribute per JanK's PR feedback

echesakovMSFT commented on src/coreclr/tools/r2rdump/CoreDisTools.cs:
```diff
@@ -24,23 +24,22 @@ public enum TargetArch
-        [DllImport(_dll)]
-        public static extern void DumpCodeBlock(IntPtr Disasm, ulong Address, IntPtr Bytes, int Size);
+        [DllImport(_dll, CallingConvention = CallingConvention.Cdecl)]
+        public static extern void DumpCodeBlock(IntPtr Disasm, IntPtr Address, IntPtr Bytes, int Size);
```
`int Size` here corresponds to `size_t Size` in https://github.com/dotnet/jitutils/blob/d9d36de5fb445662b56e583cf00746969f1bfab3/src/coredistools/coredistools.cpp#L816

Shouldn't we update that one as well?

trylek replied:
Thanks Egor for pointing that out, fixed in 3rd commit.

trylek committed:
Fix Size to be IntPtr too per Egor's PR feedback

cshung reviewed (approved):
LGTM

trylek committed:
Fix Crossgen2 build of SPC with x86-hosted R2RDump

With the change to make R2RDump architecture-specific, we need to
run it directly through the r2rdump.exe, not via dotnet which
uses the 64-bit host and can no longer run the 32-bit app.

Thanks

Tomas

jkotas commented on src/coreclr/crossgen-corelib.proj:
```diff
@@ -75,7 +75,7 @@
     
     <PropertyGroup Condition="$(BuildPdb)">
-      <CrossGenPdbCmd>$(DotNetCli) $(BinDir)\r2rdump\r2rdump.dll</CrossGenPdbCmd>
+      <CrossGenPdbCmd>$(BinDir)\r2rdump\r2rdump.exe</CrossGenPdbCmd>
```
Is this going to use some random .NET runtime installed on the machine? That does not sound good.

trylek replied:
Hmm, I don't see any trivial way to fix this as part of this PR. I'm going to revert this last commit and I'll tackle this as part of the switch-over of SPC to use CG2, probably by prioritizing my in progress change to enable building the PDB directly from Crossgen2, that should remove the need to run r2rdump from CoreCLR build.

trylek replied:
Reverted in 5th commit.

trylek committed:
Roll back the change to crossgen_corelib.proj per JanK's PR feedback

Jan Kotas expressed concerns about running R2RDump using an
arbitrary machine-wide .NET host; this isn't trivial to fix
as the runtime repo today only installs the 64-bit .NET SDK
on x64 Windows machines, I suspect that requiring explicit
32-bit .NET host would imply substantial changes to the
overall build infrastructure.

As this bit is mostly orthogonal to this change as such, I have
reverted it for now and I'll follow up as part of my ongoing task
of switching over the runtime repo to start using Crossgen2 for
producing native code.

Thanks

Tomas
